### PR TITLE
fix(decouple nipt upload from cgstats)

### PR DESCRIPTION
**Pull request by ChrOertlin**
## Description

This PR aims to decouple nipt from CGStats and retrieve relevant data from the new sample lane sequencing metrics table instead. The changes outlined below show the necessary implementation. https://github.com/Clinical-Genomics/project-planning/issues/482, https://github.com/Clinical-Genomics/cg/issues/2097

**File: `cg/meta/upload/nipt/models.py`**
- Added a new class `FlowCellQ30AndReads` to represent a summary of a flow cell.
- `FlowCellQ30AndReads` has attributes `reads` (int) and `q30` (float).
- Added methods `reads_above_threshold` and `q30_above_threshold` to `FlowCellQ30AndReads`.

**File: `cg/meta/upload/nipt/nipt.py`**
- Removed the import statement for `StatsAPI`.
- Added import statement for `FlowCellQ30AndReads` from `cg.meta.upload.nipt.models`.
- Modified the method `flowcell_passed_qc_value` to use `status_db` methods instead of `StatsAPI`.

**File: `cg/store/api/add.py`**
- Added a new method `add_sample_lane_sequencing_metrics` to create and return a new `SampleLaneSequencingMetrics` record.

**File: `cg/store/api/find_business_data.py`**
- Added methods to retrieve information from the `SampleLaneSequencingMetrics` table:
  - `get_number_of_reads_for_sample_from_sample_lane_metrics`
  - `get_number_of_reads_for_sample_on_flow_cell_from_sample_lane_metrics`
  - `get_samples_internal_ids_on_flow_cell_from_sample_lane_metrics`
  - `get_average_passing_q30_for_sample_from_sample_lane_metrics`
- Added import statement for `StatusHandler` from `cg.store.api.status`.

**File: `tests/conftest.py`**
- Added fixtures:
  - `expected_average_q30`: Returns the expected average Q30 value.
  - `expected_average_q30_on_flow_cell`: Returns the expected average Q30 value for a flow cell.
  - `sample_2`: Returns a sample name.
  - `flow_cell_2`: Returns a flow cell name.
  - `expected_total_reads_flow_cell_2`: Returns the expected total read count for a flow cell.
- Modified fixture `store_with_sequencing_metrics`:
  - Added store creation and session management.
  - Replaced direct creation of `SampleLaneSequencingMetrics` with `helpers.add_sample_lane_sequencing_metrics`.
  - Dropped the store at the end.

**File: `tests/meta/demultiplex/test_demux_post_processing.py`**
- Added import statement for `SampleLaneSequencingMetrics`.
- Modified the test function `test_add_single_sequencing_metrics_entry_to_statusdb`.

**File: `tests/meta/upload/nipt/conftest.py`**
- Added fixtures:
  - `nipt_upload_api_failed_fc_context`: Modified the fixture to use `store_failing_sequencing_qc` for `status_db`.
- Removed code related to the `stats_api` in the fixture `nipt_upload_api_failed_fc_context`.

**File: `tests/meta/upload/nipt/test_nipt_upload_api.py`**
- Removed the import statement for `NiptUploadAPI`.
- Modified the test functions `test_flowcell_passed_qc_value` and `test_flowcell_failed_qc_value`.
- Removed the unused import statement.
- Modified the test functions to use `nipt_upload_api_failed_fc_context` fixture.

**File: `tests/store/api/conftest.py`**
- Modified fixture `fixture_get_sample_by_internal_id`.

### Added

-

### Changed

-

### Fixed

-

### How to prepare for test

- [ ] Ssh to relevant server (depending on type of change)
- [ ] Use stage: `us`
- [ ] Paxa the environment: `paxa`
- [ ] Install on stage (example for Hasta):
    ```shell
    bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_main -t cg -b [THIS-BRANCH-NAME] -a
    ```

### How to test

- [ ] Do ...

### Expected test outcome

- [ ] Check that ...
- [ ] Take a screenshot and attach or copy/paste the output.

## Review

- [ ] Tests executed by
- [ ] "Merge and deploy" approved by
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a

- [ ] **MAJOR** - when you make incompatible API changes
- [ ] **MINOR** - when you add functionality in a backwards compatible manner
- [ ] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

## Implementation Plan

- [ ] Document in ...
- [ ] Deploy this branch on ...
- [ ] Inform to ...

### CODE REVIEW

File: cg/store/api/find_business_data.py
Comment (karlnyr) [inline cg/store/api/find_business_data.py:376]: This probably won't be correct, right? I think you need to do a new filter function that filters for both `sample.internal_id` and `flowcell.name`. If you do this on a sample that has been on multiple flow cells you would return the average over all flow cells.

File: cg/meta/upload/nipt/nipt.py
Comment (karlnyr) [inline cg/meta/upload/nipt/nipt.py:51]: type hint return

File: cg/meta/upload/nipt/nipt.py
Comment (karlnyr) [inline cg/meta/upload/nipt/nipt.py:63]:
```suggestion
        return (
            flow_cell_summary.q30_above_threshold(threshold=q30_threshold) and
            flow_cell_summary.reads_above_threshold(
                threshold=self.status_db.get_ready_made_library_expected_reads(case_id=case_id)
            )
```

Comment (karlnyr) [reply cg/meta/upload/nipt/nipt.py:63]: If both returns are true the return is true, if one is false its false

Comment (karlnyr) [reply cg/meta/upload/nipt/nipt.py:63]: Now that I think about it - Was there ever a log message for when one of the values failed? I really think we should log it because we would like to know if we skip the upload for some reason.

File: tests/conftest.py
Comment (karlnyr) [inline tests/conftest.py:2288]: This is great that we have a store with the new table - is there a reason we don't add it to the default store fixture instead?

File: cg/store/filters/status_metrics_filters.py
Comment (karlnyr) [inline cg/store/filters/status_metrics_filters.py:26]:
```suggestion
def filter_sample_lane_metrics_by_flow_cell_name(metrics: Query, flow_cell_name: str, **kwargs) -> Query:
```

File: cg/store/filters/status_metrics_filters.py
Comment (karlnyr) [inline cg/store/filters/status_metrics_filters.py:32]:
```suggestion
def filter_sample_lane_metrics_by_sample_internal_id(
```

File: cg/store/filters/status_metrics_filters.py
Comment (karlnyr) [inline cg/store/filters/status_metrics_filters.py:46]:
```suggestion
    FILTER_BY_FLOW_CELL_NAME: Callable = filter_sample_lane_metrics_by_flow_cell_name
    FILTER_BY_SAMPLE_INTERNAL_ID: Callable = filter_sample_lane_metrics_by_sample_internal_id
```

File: cg/store/filters/status_metrics_filters.py
Comment (karlnyr) [inline cg/store/filters/status_metrics_filters.py:1]: Is there a reason why we won't be verbose in these filter functions and file name? Metrics is a very generic name - and these filters specifically apply to the SampleLaneSequencingMetrics class.

File: tests/store/filters/test_status_metrics_filters.py
Comment (karlnyr) [inline tests/store/filters/test_status_metrics_filters.py:1]: cleeeeeean ⭐

**Review by karlnyr — CHANGES_REQUESTED** [review]
I have some comments that I do believe needs to be addressed before these changes are merged. Regarding the naming of functions etc, in my opinion, there is little gain from using abbreviate naming - but I may be wrong.

File: cg/store/api/find_business_data.py
Comment (ChrOertlin) [reply cg/store/api/find_business_data.py:376]: you are very right with this!

File: cg/meta/upload/nipt/nipt.py
Comment (ChrOertlin) [reply cg/meta/upload/nipt/nipt.py:63]: I think this is what it was before also, right? Both need to be TRUE for the QC to pass I imagined. Can you confirm?

File: cg/store/filters/status_metrics_filters.py
Comment (ChrOertlin) [reply cg/store/filters/status_metrics_filters.py:1]: they are called as SampleLaneSequencingMetricsFilter.FILTER_BY... so there is verbosity in the name already.

**Comment by seallard** [comment]
It would be nice with a more detailed PR description, maybe mentioning the ticket this PR is related to. I guess this PR is blocked until we have started populating the sequencing metrics table in parallel with cgstats in production?

**Comment by ChrOertlin** [comment]
> It would be nice with a more detailed PR description, maybe mentioning the ticket this PR is related to. I guess this PR is blocked until we have started populating the sequencing metrics table in parallel with cgstats in production?

Yes, until the samplelanesequencing metrics is up and running this is blocked

File: cg/store/api/find_business_data.py
Comment (Vince-janv) [inline cg/store/api/find_business_data.py:329]: Is this used anywhere in the PR? If not I think it should be added when used

File: cg/store/api/find_business_data.py
Comment (Vince-janv) [inline cg/store/api/find_business_data.py:360]: Clean ✨

File: cg/store/api/find_business_data.py
Comment (Vince-janv) [inline cg/store/api/find_business_data.py:372]: Do we want to return strings or Sample objects? I feel like such a function would be more flexible?

File: cg/store/api/find_business_data.py
Comment (Vince-janv) [inline cg/store/api/find_business_data.py:382]:
```suggestion
        sequencing_metrics: List[SampleLaneSequencingMetrics] = apply_metrics_filter(
```

File: cg/store/api/find_business_data.py
Comment (Vince-janv) [inline cg/store/api/find_business_data.py:388]: What does `total_passing_q30` mean?

File: cg/store/api/find_business_data.py
Comment (Vince-janv) [inline cg/store/api/find_business_data.py:390]: I feel like this should be weighted by the number of reads on the lane, to be accurate

File: cg/store/api/find_business_data.py
Comment (Vince-janv) [inline cg/store/api/find_business_data.py:410]: We only need one database interaction here right? Fetch the sequencing_metrics for the flow cell and then sum the reads?

File: tests/store/api/conftest.py
Comment (Vince-janv) [inline tests/store/api/conftest.py:185]: I think you can use `base_store` here to not have to create it on your own

File: tests/conftest.py
Comment (Vince-janv) [inline tests/conftest.py:2695]: are there not two flow cell fixtures already?

File: tests/conftest.py
Comment (Vince-janv) [inline tests/conftest.py:2689]: Use `mother_sample_id` or `father_sample_id` instead

**Review by Vince-janv — COMMENTED** [review]
Oof nice untangling of cgstats from this module 🪢  Some comments about the querying. I think the number of queries can be reduced by returning the models in some places instead of only strings

File: cg/store/api/find_business_data.py
Comment (ChrOertlin) [reply cg/store/api/find_business_data.py:329]: Apparently not used - I removed it and associated tests. Nothing seems to break because of it. It was not used anywhere in the code

File: cg/store/api/find_business_data.py
Comment (ChrOertlin) [reply cg/store/api/find_business_data.py:390]: It is a percentage of bases that passes a filter. We sum the percentages here and then get the average when we return

File: cg/store/api/find_business_data.py
Comment (ChrOertlin) [reply cg/store/api/find_business_data.py]: yes, I simplified these functions by a lot now.

File: cg/store/api/find_business_data.py
Comment (Vince-janv) [reply cg/store/api/find_business_data.py]: Yes, but should the average not be weighted? If we have two lanes one with 10 reads and q30 of 50 and the other lane has 1000 reads and a q30 of 90, then the average should not be 70 right? Rather something like (10*50+1000*90)/(10+1000)

Comment (ChrOertlin) [reply cg/store/api/find_business_data.py]: it is a measure on base level not reads, so you would want normalise to total bases per lane?

Comment (Vince-janv) [reply cg/store/api/find_business_data.py]: Does that not make sense? I feel like the number of bases will always be proportional to the number of reads. And if we get much more data from one lane then the quality of that lane should weigh more than the results from other lanes

Comment (ChrOertlin) [reply cg/store/api/find_business_data.py]: just a scenario I could think of:
lane 1 low reads, abmysmal q30.
lane 2 high reads, good q30 
Weighted average passes QC threshold which causes poor data to be delivered. 

I think this suggestion would change the current implementation (taken an average) as a cutoff. Maybe we can make an issue and discuss this in a refinement meeting?

Comment (Vince-janv) [reply cg/store/api/find_business_data.py]: Issue created: https://github.com/Clinical-Genomics/cg/issues/2285

File: cg/store/api/add.py
Comment (ChrOertlin) [reply cg/store/api/add.py:379]: We use this somewhere in the demux api as well, I am planning on moving some functionality we made there to the store this is one of it. I can remove it here now and add it later with new implementation

Comment (ChrOertlin) [reply cg/store/api/add.py:379]: Ahwell I actually use it in the store helpers in this PR so I will keep it here as well

File: tests/conftest.py
Comment (ChrOertlin) [reply tests/conftest.py]: fixed

File: tests/conftest.py
Comment (ChrOertlin) [reply tests/conftest.py]: fixed

File: tests/store/api/conftest.py
Comment (ChrOertlin) [reply tests/store/api/conftest.py:185]: fixed

**Comment by ChrOertlin** [comment]
# Test cg upload nipt case rarepony

Failure due to existing statina entry, DEBUG log shows passed QC for case

image

File: cg/store/api/find_business_data.py
Comment (Vince-janv) [inline cg/store/api/find_business_data.py:335]: Docstring

File: cg/store/api/find_business_data.py
Comment (Vince-janv) [inline cg/store/api/find_business_data.py:353]: More verbose docstring
```suggestion
        """Calculates the average q30 for each sample on a flow cell and returns the average between the samples."""
```

File: cg/store/api/find_business_data.py
Comment (Vince-janv) [inline cg/store/api/find_business_data.py:359]: Make it a set directly
```suggestion
        unique_sample_internal_ids: Set[str] =
            {sequencing_metric.sample_internal_id for sequencing_metric in sequencing_metrics}
```

File: cg/meta/upload/nipt/models.py
Comment (Vince-janv) [inline cg/meta/upload/nipt/models.py:19]: I think this sounds more like more like a bool return
```suggestion
    def passes_read_threshold(self, threshold: int) -> bool:
        """Check if the number of reads is above a threshold."""
        return self.total_reads_on_flow_cell >= threshold
```

File: cg/meta/upload/nipt/models.py
Comment (Vince-janv) [inline cg/meta/upload/nipt/models.py:23]: Same here
```suggestion
    def passes_q30_threshold(self, threshold: float) -> bool:
        """Check if the Q30 is above a threshold."""
        return self.average_q30_across_samples >= threshold
```

**Review by Vince-janv — APPROVED** [review]
Well done Chris 👍

File: cg/store/api/find_business_data.py
Comment (seallard) [inline cg/store/api/find_business_data.py:371]: What if the `unique_sample_internal_ids` is empty?

Comment (seallard) [reply cg/store/api/find_business_data.py:371]: Might be enough to add `if sum_average_q30_across_samples and len(unique_sample_internal_ids)`

File: cg/store/api/find_business_data.py
Comment (seallard) [inline cg/store/api/find_business_data.py:386]: I think the `if reads_count else 0` is unnecessary since `reads_count` is initialised to 0, right?

File: cg/meta/upload/nipt/nipt.py
Comment (seallard) [inline cg/meta/upload/nipt/nipt.py:58]: A matter of taste, but I don't think the caller needs to know which table in the database the information is retrieved from. I would say that is an implementation detail of the function and it is not something we want the caller to think about. The caller is only interested in that the function returns the number of reads associated with a flow cell, how it is done is not of interest to the caller and is just an implementation detail.

So (in my opinion) names like these would be better because they don't leak implementation details
```python
get_average_fraction_passing_q30_for_flow_cell(flow_cell.name)
get_number_of_reads_for_flow_cell(flow_cell.name)
```

**Review by seallard — APPROVED** [review]
Nice 👍 I think some of the function names were needlessly long, but feel free to disregard if my reasoning does not make sense.

File: cg/store/api/find_business_data.py
Comment (ChrOertlin) [reply cg/store/api/find_business_data.py:386]: yes that is true, fixed it

File: cg/store/api/find_business_data.py
Comment (ChrOertlin) [reply cg/store/api/find_business_data.py:371]: I added a check for this!

File: cg/store/api/add.py
Comment (ChrOertlin) [reply cg/store/api/add.py:372]: It is for the non-required fields when you initialise a model.

File: cg/meta/upload/nipt/nipt.py
Comment (ChrOertlin) [reply cg/meta/upload/nipt/nipt.py:58]: I shortened the function names!

**Review by karlnyr — APPROVED** [review]
Excellent - the calculation seems correct now - wondering if we should return 0 or do some kind of exception raise when we have the weird numbers in `get_average_fraction_passing_q30_for_flow_cell` but other than that I am very satisfied - nice work :)

**Comment by ChrOertlin** [comment]
Stage:
Logging deploy ...
Getting deployer... done.
Getting last commit message and SHA... done.
Getting version of deploy scripts... /home/proj/stage/demultiplexed-runs
done.
Log deploy... done.
cg, version 40.6.1

Prod:
Logging deploy ...
Getting deployer... done.
Getting last commit message and SHA... done.
Getting version of deploy scripts... /home/proj/stage/demultiplexed-runs
done.
Log deploy... done.
cg, version 40.6.1

**Pull request merged**
